[XLA:CPU][oneDNN] Enable oneDNN Convolution Custom Calls in Thunk Runtime

othakkar opened the pull request:
This PR enables support for oneDNN Convolution operations in the XLA:CPU Thunk runtime, building upon the foundational implementation of `OneDnnOpThunk`.

Key changes:
- Updated thunk emitter to emit `OneDnnOpThunk` for oneDNN convolution during compilation.
- Enabled oneDNN custom call rewrite for Convolution in `onednn_contraction_rewriter.cc`. 
- Added support for oneDNN Convolution op via `ExecuteOneDnnConvolution(...)` in `onednn_convolution.cc`.

othakkar committed:
enable oneDNN convolution in thunk runtime

othakkar commented:
@penpornk, please review this PR when you have a chance. Thanks!

othakkar committed:
add a test for oneDNN conv thunk

penpornk reviewed (changes requested):
Thank you for the PR and sorry for the delay!

othakkar committed:
Merge branch 'main' into othakkar/onednn_thunk_conv_nonFFI

othakkar committed:
addressing review comments

othakkar commented on xla/backends/cpu/runtime/onednn/onednn_op_thunk_test.cc:
```diff
@@ -106,6 +106,124 @@ TEST(OneDnnOpThunkTest, SimpleOneDnnMatMulThunk) {
 }
 
+// A very small 2D NHWC convolution test exercising the oneDNN convolution path
+// through OneDnnOpThunk. We build a 1x1 kernel with weight=2 so the expected
```
Done (https://github.com/openxla/xla/pull/31511/commits/c103dd964d6a59319c30c9aabb4345dfa4a882c7).

othakkar commented on xla/backends/cpu/runtime/onednn/onednn_op_thunk_test.cc:
```diff
@@ -106,6 +106,124 @@ TEST(OneDnnOpThunkTest, SimpleOneDnnMatMulThunk) {
+  OneDnnConvolutionConfig conv_config;
+  conv_config.set_dims(4);  // rank
+  // Input layout NHWC: batch=0, feature=3, spatial (H,W) = (1,2) zero-based.
+  auto* inp = conv_config.mutable_input();
```
Thanks for sharing this. Done (https://github.com/openxla/xla/pull/31511/commits/c103dd964d6a59319c30c9aabb4345dfa4a882c7).

othakkar commented on xla/backends/cpu/runtime/onednn/onednn_op_thunk_test.cc:
```diff
@@ -106,6 +106,124 @@ TEST(OneDnnOpThunkTest, SimpleOneDnnMatMulThunk) {
+  win->add_pad_right(1);
+  // Dilations store (actual +2) so 3 -> 1 actual dilation.
+  win->add_window_dilations(3);
+  win->add_window_dilations(3);
```
This encoding is not a oneDNN requirement, but an internal encoding defined in [onednn_contraction_rewriter.cc](https://github.com/openxla/xla/blob/main/xla/service/cpu/onednn_contraction_rewriter.cc#L681). This was done to reserve `0` to denote "unset" (for example, zero padding).
Updated the comments to address this better in https://github.com/openxla/xla/blob/main/xla/service/cpu/onednn_contraction_rewriter.cc#L681.

othakkar commented on xla/backends/cpu/runtime/onednn/onednn_op_thunk_test.cc:
```diff
@@ -106,6 +106,124 @@ TEST(OneDnnOpThunkTest, SimpleOneDnnMatMulThunk) {
+  auto* ker = conv_config.mutable_kernel();
+  ker->set_dims(4);
+  auto* filter = ker->mutable_filter();
+  filter->set_input_feature_dim(2);   // zero-based index of IC
+  filter->set_output_feature_dim(3);  // zero-based index of OC
+  // Spatial dims (H,W) one-based: (0->1,1->2) => 1,2
+  filter->add_spatial_dims(1);
+  filter->add_spatial_dims(2);
```
This encoding is not a oneDNN requirement, but an internal encoding defined in [onednn_contraction_rewriter.cc](https://github.com/openxla/xla/blob/main/xla/service/cpu/onednn_contraction_rewriter.cc#L681). This was done to reserve `0` to denote "unset" (for example, zero padding).
Updated the comments to address this better in https://github.com/openxla/xla/blob/main/xla/service/cpu/onednn_contraction_rewriter.cc#L681.

penpornk replied:
@akhilgoe You're right. We have recently done a major protobuf version update -- from 3.21.9 to 5.28.3 in June (https://github.com/openxla/xla/commit/1fc2a846d259d3c896a6177485abcc68d97627a9), and to v6.31.1 in August (https://github.com/openxla/xla/pull/29658/commits/7abd4055b1629815d53dff1647a3ec15b7a79dc6). So let's test if the issue was gone with the upgrade or if it still persists.

othakkar commented on xla/service/cpu/onednn_convolution.cc:
```diff
@@ -208,6 +209,147 @@ CreateOneDnnPrimDesc<dnnl::convolution_forward::primitive_desc>(
       any_res_md, strides, rhs_dilations, pad_left, pad_right, attrs);
 }
 
+void ExecuteOneDnnConvolution(absl::Span<MemrefInfoHandler> arguments,
```
I plan to submit a cleanup PR for all four oneDNN primitives to remove all dead code related to the legacy runtime, later, once the thunk runtime changes are merged. This is to avoid making this (and other thunk) PRs longer than needed. Let me know if that is okay.

othakkar committed:
clang-format checks

othakkar committed:
restore changes lost while merging with main

othakkar committed:
remove dead-code related to oneDNN convolution for legacy runtime

othakkar committed:
fixup: onednn_contraction_rewriter - lint related

othakkar committed:
replace auto with corresponding type

othakkar commented on xla/service/cpu/onednn_contraction_rewriter.cc:
```diff
@@ -554,10 +554,6 @@ bool OneDnnContractionRewriter::ShouldRewriteDot(
-  // support in thunk runtime.
-  return false;
-
   // NOLINTBEGIN(clang-diagnostic-unreachable-code)
```
Done (https://github.com/openxla/xla/pull/31511/commits/9a4ba20c1624fb3d51e3a43cb805e1f4ce7dede7)

othakkar commented:
@penpornk I added a commit (https://github.com/openxla/xla/pull/31511/commits/1bc5c7364e81e4e867d4763ae57eb1c15907d01e) to remove the code related to convolution for the legacy runtime. Please let me know if this looks okay. I addressed your other comments.

penpornk reviewed (approved):
Thank you again for the PR! Please rebase because this PR conflicts with https://github.com/openxla/xla/commit/968b761b26d600620be5f4458aa3376cb6210295 that recently went in.

othakkar committed:
Merge remote-tracking branch 'upstream/main' into othakkar/onednn_thunk_conv_nonFFI

penpornk reviewed (approved):
Thank you again!

othakkar committed:
Merge remote-tracking branch 'upstream/main' into othakkar/onednn_thunk_conv_nonFFI
